drivers/gpu/drm/panel/panel-samsung-s6e3fa0-full.c

```diff
@@ -0,0 +1,764 @@
+	} while (0)
+
+static inline
+struct s6e3fa0_ctx *panel_to_s6e3fa0(struct drm_panel *panel)
+{
+	return container_of(panel, struct s6e3fa0_ctx, panel);
+}
+
+enum s6e3fa0_test_key_level {
+	TK_LEVEL_1,
+	TK_LEVEL_2
+};
+
+static int s6e3fa0_test_key_toggle(struct s6e3fa0_ctx *ctx, bool enable,
+				   enum s6e3fa0_test_key_level level)
+{
+	struct mipi_dsi_device *dsi = ctx->dsi;
+	int ret;
+	const u8 *seq;
+
+	if (!ctx->seq_data)
+		return -ENODEV;
+
+	if (enable)
+		seq = ctx->seq_data->test_key_enable_seq;
+	else
+		seq = ctx->seq_data->test_key_disable_seq;
+
+	/* First 3 bytes - level 1 lock/unlock */
+	ret = mipi_dsi_generic_write(dsi, seq, 3);
+	if (ret < 0)
+		return ret;
+
+	/*
+	 * During init lock/unlock 2nd level MTP commands (send bytes 4,5,6)
+	 * For set_brightness - depends on panel subtype.
+	 */
+	if (level == TK_LEVEL_2 || ctx->seq_data->brightness_use_level2)
+		ret = mipi_dsi_generic_write(dsi, seq + 3, 3);
+	return ret;
+}
+
+static int s6e3fa0_write_aid_control(struct s6e3fa0_ctx *ctx, int candela)
+{
+	struct mipi_dsi_device *dsi = ctx->dsi;
+	int idx;
+	const u8 *seq;
+
+	if (!ctx->seq_data)
+		return -ENODEV;
+
+	idx = map_candela_to_aid[candela];
+	seq = (*ctx->seq_data->aid_seq)[idx];
+	return mipi_dsi_generic_write(dsi, seq, AID_SEQUENCE_LEN);
+}
+
+static int s6e3fa0_write_elvss(struct s6e3fa0_ctx *ctx, int candela)
+{
+	struct mipi_dsi_device *dsi = ctx->dsi;
+	int idx;
+	const u8 *seq;
+
+	if (!ctx->seq_data)
+		return -ENODEV;
+
+	idx = map_candela_to_elvss[candela];
+	seq = (*ctx->seq_data->elvss_seq)[idx];
+	return mipi_dsi_generic_write(dsi, seq, ELVSS_SEQUENCE_LEN);
+}
+
+static int s6e3fa0_write_gamma(struct s6e3fa0_ctx *ctx, int candela)
+{
+	struct mipi_dsi_device *dsi = ctx->dsi;
+	const u8 *seq;
+
+	if (!ctx->seq_data)
+		return -ENODEV;
+
+	/* Gamma sequence is directly indexed by candela value */
+	seq = (*ctx->seq_data->gamma_seq)[candela];
+	return mipi_dsi_generic_write(dsi, seq, GAMMA_SEQ_LEN);
+}
+
+static int s6e3fa0_write_gamma_apply(struct s6e3fa0_ctx *ctx)
+{
+	/* From vendor driver: Gamma, LTPS(AID) update */
+	dsi_generic_write_seq(ctx->dsi, 0xf7, 0x03);
+	/* S6E3FA0 has additional command */
+	if (ctx->subtype == PANEL_S6E3FA0)
+		dsi_generic_write_seq(ctx->dsi, 0xf7, 0x00);
+	return 0;
+}
+
+static int s6e3fa0_set_brightness(struct backlight_device *bldev)
+{
+	struct s6e3fa0_ctx *ctx = bl_get_data(bldev);
+	const int candela = backlight_get_brightness(bldev);
+	int r;
+
+	r = s6e3fa0_test_key_toggle(ctx, true, TK_LEVEL_1);
+	if (r < 0)
+		return r;
+	r = s6e3fa0_write_aid_control(ctx, candela);
+	if (r < 0)
+		return r;
+	r = s6e3fa0_write_elvss(ctx, candela);
+	if (r < 0)
+		return r;
+	r = s6e3fa0_write_gamma(ctx, candela);
+	if (r < 0)
+		return r;
+	r = s6e3fa0_write_gamma_apply(ctx);
+	if (r < 0)
+		return r;
+	r = s6e3fa0_test_key_toggle(ctx, false, TK_LEVEL_1);
+	if (r < 0)
+		return r;
+
+	return 0;
+}
+
+static int s6e3fa0_detect_panel_type(struct s6e3fa0_ctx *ctx)
```

Comment:
This will be not useful?

Reply:
I left this in for the purpose of verifying that the panel is actually LCD ID 0x414024, figured it would not hurt to check that the panel actually matches the one we expect even if we're not selecting between multiple panels.  Your thoughts?

Reply:
If upstream accepts it, then maybe it's fine. I don't have a strong opinion here

Reply:
Left in panel ID check but cleaned up some unused code